File: src/main/java/co/herod/adbwrapper/Adb.java
package co.herod.adbwrapper;

import co.herod.adbwrapper.model.AdbDevice;
import co.herod.adbwrapper.util.StringUtils;
import io.reactivex.Observable;
import io.reactivex.Single;

import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Created by matthewherod on 23/04/2017.
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public class Adb {

    public static Observable<AdbDevice> devices() {

        return ProcessHelper.observableProcess(AdbProcesses.devices())
                .filter(StringUtils::containsTab)
                .map(AdbDevice::parseAdbString);
    }

    public static void typeText(final AdbDevice adbDevice, final String inputText) {
        ProcessHelper.blocking(AdbProcesses.inputText(adbDevice, inputText), 5, TimeUnit.SECONDS);
    }

    public static void pressKeyBlocking(final AdbDevice adbDevice, final int key) {
        ProcessHelper.blocking(AdbProcesses.pressKey(adbDevice, key), 5, TimeUnit.SECONDS);
    }

    public static Observable<Map<String, String>> getDisplayDumpsys(final AdbDevice adbDevice) {
        return dumpsysMap(adbDevice, AdbDeviceProperties.PROPS_DISPLAY).toObservable();
    }

    public static Observable<Map<String, String>> getInputMethodDumpsys(final AdbDevice adbDevice) {
        return dumpsysMap(adbDevice, AdbDeviceProperties.PROPS_INPUT_METHOD).toObservable();
    }

    public static Single<Map<String, String>> dumpsysMap(final AdbDevice adbDevice, final String type) {

        return ProcessHelper.observableProcess(AdbProcesses.dumpsys(adbDevice, type))
                .filter(StringUtils::containsKeyValueSeparator)
                .map(StringUtils::splitKeyValue)
                .toMap(s -> s[0].trim(), s -> s[1].trim());
    }

    public static Observable<String> dumpUiHierarchy(final AdbDevice adbDevice) {
        return ProcessHelper.observableProcess(AdbProcesses.dumpUiHierarchyProcess(adbDevice));
    }

    static void blocking(final AdbDevice device, final String command) {
        ProcessHelper.blocking(AdbProcesses.adb(device, command));
    }

    public static Observable<String> command(final AdbDevice adbDevice, final String command) {
        return ProcessHelper.observableProcess(AdbProcesses.adb(adbDevice, command));
    }
}

File: src/main/java/co/herod/adbwrapper/Adb.kt
package co.herod.adbwrapper

import co.herod.adbwrapper.model.AdbDevice
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single
import java.util.concurrent.TimeUnit

/**
 * Created by matthewherod on 23/04/2017.
 */
object Adb {

    val processFactory: IProcessFactory = ProcessFactory()

    const val DEVICES = "devices"

    fun devices(): Observable<AdbDevice> = processFactory.observableProcess(AdbProcesses.devices())
            .filter { "\t" in it }
            .map { AdbDevice.parseAdbString(it) }

    fun typeText(adbDevice: AdbDevice, inputText: String): ProcessBuilder? =
            AdbProcesses.inputText(adbDevice, inputText).also {
                it?.let { processFactory.blocking(it, 5, TimeUnit.SECONDS) }
            }

    fun pressKeyBlocking(adbDevice: AdbDevice?, key: Int) =
            AdbProcesses.pressKey(adbDevice, key)?.let {
                processFactory.blocking(it, 5, TimeUnit.SECONDS)
            }

    fun getDisplayDumpsys(adbDevice: AdbDevice): Flowable<Map<String, String>> =
            adbDevice.dumpsysMap(AdbDeviceProperties.PROPS_DISPLAY).toFlowable()

    fun getInputMethodDumpsys(adbDevice: AdbDevice): Flowable<Map<String, String>> =
            adbDevice.dumpsysMap(AdbDeviceProperties.PROPS_INPUT_METHOD).toFlowable()

    private fun AdbDevice.dumpsysMap(type: String): Single<Map<String, String>> =
            processFactory.observableProcess(AdbProcesses.dumpsys(this, type))
                    .filter { "=" in it }
                    .map { it.trim { it <= ' ' }.split("=".toRegex(), 2) }
                    .toMap({ it[0].trim { it <= ' ' } }) { it[1].trim { it <= ' ' } }

    fun dumpUiHierarchy(adbDevice: AdbDevice?): Observable<String> =
            processFactory.observableProcess(AdbProcesses.dumpUiHierarchyProcess(adbDevice))

    fun command(adbDevice: AdbDevice, command: String): Observable<String> =
            processFactory.observableProcess(AdbProcesses.adb(adbDevice, command))

    internal fun blocking(device: AdbDevice?, command: String) {
        processFactory.blocking(AdbProcesses.adb(device, command))
    }
}
